### ISSUE TITLE
Reopening Issue#70 - Failed to execute: aa2nonred.pl 

### ISSUE THREAD

**Issue by psur9757** [Gaius-Augustus/BRAKER#75]
I have updated BRAKER and Augustus.  I had the HPC team update everything as instructed.

Also, I give the PBS script diamond but it still runs blast+ and gives the same error.

PBS script
```
#!/bin/bash
#PBS -P OSR
#PBS -N brakerP1
#PBS -l select=1:ncpus=8:mem=64GB
#PBS -l walltime=20:00:00
#PBS -e /scratch/OSR/OSR409/results/annotation/structural/brakerP1/braker.err
#PBS -o /scratch/OSR/OSR409/results/annotation/structural/brakerP1/braker.out

wdir=/scratch/OSR/OSR409/results/annotation
outdir=$wdir/structural/brakerP1
indir=$wdir/structural/star

module load genemark-es/4.46
module load augustus/3.3.2
module load bamtools/2.5.1
module load blast+/2.7.1
module load samtools/1.9
module load python/3.7.2
module load genomethreader/1.7.1
module load makehub/i
module load eval/2.2.8
module load braker2/2.1.4
module load diamond/0.9.25
module load cdbfasta/1.00

export AUGUSTUS_CONFIG_PATH=$outdir/augustus_config
export AUGUSTUS_BIN_PATH=/usr/local/augustus/3.3.2/bin
export AUGUSTUS_SCRIPTS_PATH=/usr/local/augustus/3.3.2/scripts

braker.pl --species=OSR --genome=$wdir/ref.fa --bam=$indir/refAligned.sortedByCoord.out.bam --prot_seq=$wdir/prot.fa --prg=gth --gff3 --softmasking --UTR=on --AUGUSTUS_ab_initio --cores=8 --fungus --crf --workingdir=$outdir 
```
braker.log
```
$ tail braker.log 
# Thu Sep 12 02:48:39 2019: fixing utrrnaseq output

mv /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utrs.f.gff /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utrs.gff
# Thu Sep 12 02:48:39 2019: Creating gb file for UTR training

cat /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utrs.gff /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/augustus.hints.f.gtf | grep -P "(CDS|5'-UTR|3'-UTR)" | sort -n -k 4,4 | sort -s -k 10,10 | sort -s -k 1,1 >> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/genes.gtf 2> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/errors/cat_utrs_augustus_noUtrs.err

perl /usr/local/augustus/3.3.2/scripts/gff2gbSmallDNA.pl /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/genes.gtf /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/genome.fa 1074 /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr.gb --good=/scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/bothutr.lst 1> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/gff2gbSmallDNA.utr.stdout 2> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/errors/gff2gbSmallDNA.utr.stderr
# Thu Sep 12 02:48:40 2019: BLAST training gene structures (with UTRs) against themselves:
perl /usr/local/augustus/3.3.2/scripts/aa2nonred.pl /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.fa /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.nr.fa --BLAST_PATH= --cores=8 1> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr.aa2nonred.stdout 2> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/errors/utr.aa2nonred.stderr
```
braker.err
```
	The file ~/.gmkey exists and has not been copied.
Use of uninitialized value $BLAST_PATH in concatenation (.) or string at /usr/local/braker2/2.1.4/scripts/braker.pl line 9685.
ERROR in file /usr/local/braker2/2.1.4/scripts/braker.pl at line 9697
Failed to execute: perl /usr/local/augustus/3.3.2/scripts/aa2nonred.pl /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.fa /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.nr.fa --BLAST_PATH= --cores=8 1> /s
cratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr.aa2nonred.stdout 2> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/errors/utr.aa2nonred.stderr!
```
errors/utr.aa2nonred.stderr
```
Option blast_path requires an argument
BLAST options error: File /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.fa.nonreddb is empty
ERROR in file /usr/local/augustus/3.3.2/scripts/aa2nonred.pl at line 220
Could not open /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.fa.blastout!
```
utr.aa2nonred.stdout
```
# Thu Sep 12 02:48:40 2019: Did not find environment variable $BLAST_PATH
# Thu Sep 12 02:48:40 2019: Trying to guess $BLAST_PATH from location of blastp executable that is available in your $PATH.
# Thu Sep 12 02:48:40 2019: Setting $BLAST_PATH to /usr/local/blast+/2.7.1/bin


Building a new DB, current time: 09/12/2019 02:48:47
New DB name:   /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.fa.nonreddb
New DB title:  /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.fa.nonreddb
Sequence type: Protein
Keep MBits: T
Maximum file size: 1000000000B
```

https://github.com/Gaius-Augustus/BRAKER/issues/70#issuecomment-533925545

**Comment by KatharinaHoff** [comment]
Thank you for this detailed report. I hope commit https://github.com/Gaius-Augustus/BRAKER/commit/81f6b9b41e4c824239292cdf48c777a4d27a0bfe will solve this problem. (At least the point where braker.pl did not use DIAMOND during filtering UTR training examples for redundancy. UTR training may have other issues...)

You don't have to re-run BRAKER from scratch on your data after updating the code (you need to get it with 'git clone https://github.com/Gaius-Augustus/BRAKER.git' since I did not make a new release, yet). If you have stored the AUGUSTUS parameters from the previous run, try the following:

Check in the log file of your old BRAKER run what was the size of flanking region (there is a line that contains "gff2gbSmallDNA.pl"; the command in that line has an integer parameter after the genome file, e.g. 1750 or similar; use that number instead of 1750 in the following):
```

FLANK_SIZE=1750 # adapt!

braker.pl --species=OSR --genome=$wdir/ref.fa --bam=$indir/refAligned.sortedByCoord.out.bam --prot_seq=$wdir/prot.fa --prg=gth --gff3 --softmasking --UTR=on --AUGUSTUS_ab_initio --cores=8 --fungus --crf --workingdir=$outdir --useexisting --AUGUSTUS_hints_preds=/scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/augustus.hints.gtf  --flanking_DNA=${FLANK_SIZE} --workingdir=OSR_utr
```
Good luck :-)

Katharina

**Comment by psur9757** [comment]
Hi @KatharinaHoff, unfortunately there is a new error at the same step. The HPC team updated everything and I started BRAKER from scratch just to be sure. Below are the relevant files.

1. standard error file
```
$ tail braker.err 
	The file ~/.gmkey exists and has not been copied.
ERROR in file /usr/local/braker2/2.1.4a/scripts/braker.pl at line 9721
Failed to execute: perl /usr/local/augustus/3.3.2/scripts/aa2nonred.pl /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.fa /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.nr.fa --DIAMOND_PATH=/usr/local/diamond/0.9.25/bin --cores=8 --diamond 1> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr.aa2nonred.stdout 2> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/errors/utr.aa2nonred.stderr!
```
2. standard output file
```
$ tail braker.out
# Tue Oct  1 09:23:42 2019: Log information is stored in file /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/braker.log
```

3. BRAKER log
```
$ tail braker.log 
# Tue Oct  1 15:44:06 2019: fixing utrrnaseq output

mv /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utrs.f.gff /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utrs.gff
# Tue Oct  1 15:44:06 2019: Creating gb file for UTR training

cat /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utrs.gff /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/augustus.hints.f.gtf | grep -P "(CDS|5'-UTR|3'-UTR)" | sort -n -k 4,4 | sort -s -k 10,10 | sort -s -k 1,1 >> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/genes.gtf 2> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/errors/cat_utrs_augustus_noUtrs.err

perl /usr/local/augustus/3.3.2/scripts/gff2gbSmallDNA.pl /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/genes.gtf /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/genome.fa 1074 /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr.gb --good=/scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/bothutr.lst 1> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/gff2gbSmallDNA.utr.stdout 2> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/errors/gff2gbSmallDNA.utr.stderr
# Tue Oct  1 15:44:06 2019: DIAMOND/BLAST training gene structures (with UTRs) against themselves:
perl /usr/local/augustus/3.3.2/scripts/aa2nonred.pl /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.fa /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.nr.fa --DIAMOND_PATH=/usr/local/diamond/0.9.25/bin --cores=8 --diamond 1> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr.aa2nonred.stdout 2> /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/errors/utr.aa2nonred.stderr
```

4. aa2nonred standard error file
```
$ tail errors/utr.aa2nonred.stderr
Licensed under the GNU GPL <https://www.gnu.org/licenses/gpl.txt>
Check http://github.com/bbuchfink/diamond for updates.

#CPU threads: 8
Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)
Temporary directory: /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1
Opening the database...  [0.084472s]
Error: Incomplete database file. Database building did not complete successfully.
ERROR in file /usr/local/augustus/3.3.2/scripts/aa2nonred.pl at line 220
Could not open /scratch/RDS-FAE-OSR-RW/OSR409/results/annotation/structural/brakerP1/utr_genes_in_gb.fa.blastout!
```

5. aa2nonred standard output file
```
$ tail utr.aa2nonred.stdout
# Tue Oct  1 15:44:07 2019: Did not find environment variable $DIAMOND_PATH
# Tue Oct  1 15:44:07 2019: Setting $DIAMOND_PATH to command line argument --DIAMOND_PATH value /usr/local/diamond/0.9.25/bin.
```

**Comment by MartijnHolterman** [comment]
Hi all, I just ran into the same problem and I may have found a solution. 
When looking at the error files I noticed that rnaseq2utr_minus.err and rnaseq2utr_plus.err were absolutely massive. So after some digging it turns out that the bamToWig.py script generates a tab-delimited .wig file, but the next program in line, utrrnaseq, expects a space-delimited .wig file as input (that's also the official format I think).
Changing line 263 in the bamToWig.py script to: 
```
wig_handle.write(start` + " " + depth + "\n")
```
solved the issue for me. I'm now running into a problem with the intron annotation, but that's a different issue ;-)
